tempesta_fw/http.c

```diff
@@ -1153,28 +1173,32 @@ tfw_http_resp_cache_cb(TfwHttpReq *req, TfwHttpResp *resp)
  * Request messages that were forwarded to a backend server are added
  * to and kept in @msg_queue of the connection @conn for that server.
  * If a paired request is not found, then the response is deleted.
+ *
+ * If thre isn't paired client request, then it seems upsream server is
+ * misbehaving, so the caller has to drop the server connection.
  */
 static TfwHttpReq *
 tfw_http_popreq(TfwHttpMsg *hmresp)
 {
-	TfwMsg *msg;
+	TfwHttpReq *req;
 	TfwConnection *conn = hmresp->conn;
 
 	spin_lock(&conn->msg_qlock);
-	if (unlikely(list_empty(&conn->msg_queue))) {
+
+	req = list_first_entry_or_null(&conn->msg_queue, TfwHttpReq,
+				       msg.msg_list);
+	if (unlikely(!req || !(req->flags & TFW_HTTP_MSG_SENT))) {
```

Comment:
I'd say that a (write) memory barrier is justified before checking for `TFW_HTTP_MSG_SENT` flag. That would ensure that all writes are committed before the check is performed.